bleachbit/Cleaner.py

```diff
@@ -370,16 +370,16 @@ def get_commands(self, option_id):
                         'custom folder has invalid type %s' % c_type)
 
         # menu
-        menu_dirs = ['~/.local/share/applications',
-                     '~/.config/autostart',
-                     '~/.gnome/apps/',
-                     '~/.gnome2/panel2.d/default/launchers',
-                     '~/.gnome2/vfolders/applications/',
-                     '~/.kde/share/apps/RecentDocuments/',
-                     '~/.kde/share/mimelnk',
-                     '~/.kde/share/mimelnk/application/ram.desktop',
-                     '~/.kde2/share/mimelnk/application/',
-                     '~/.kde2/share/applnk']
+        menu_dirs = (os.path.expanduser('~/.local/share/applications'),
+                     os.path.expanduser('~/.config/autostart'),
+                     os.path.expanduser('~/.gnome/apps/'),
+                     os.path.expanduser('~/.gnome2/panel2.d/default/launchers'),
+                     os.path.expanduser('~/.gnome2/vfolders/applications/'),
+                     os.path.expanduser('~/.kde/share/apps/RecentDocuments/'),
+                     os.path.expanduser('~/.kde/share/mimelnk'),
+                     os.path.expanduser('~/.kde/share/mimelnk/application/ram.desktop'),
+                     os.path.expanduser('~/.kde2/share/mimelnk/application/'),
+                     os.path.expanduser('~/.kde2/share/applnk'))
 
         if 'posix' == os.name and 'desktop_entry' == option_id:
             for dirname in menu_dirs:
```

Comment:
Since the nested for-loop is already using a list comprehension, could I just do following?
```python
for dirname in menu_dirs:
    for filename in [fn for fn in children_in_directory(os.path.expanduser(dirname), False)
                     if fn.endswith('.desktop')]:
```
Or even better, could I also use a generator expression to take advantage of lazy evaluation while assigning it to variable to enhance readability and apply it to the nested loop?
```python
for dirname in menu_dirs:
    filenames = (
        fn for fn in children_in_directory(os.path.expanduser(dirname), False)
        if fn.endswith('.desktop'))
    for filename in filenames:
```
Yeah, I know it's a refactoring. It's my specialty and hobby. 😎

Reply:
You can completely remove the comprehension to make the code more readable.

Reply:
Good point! Pushing change now.

Reply:
LGTM.

I usually don't replace lists with tuples, because it affects readability. With `[something]` it is always clear that `something` is a list item. While `(something)` may also look like a param to some function call.

Reply:
Another reason why I use lists instead of tuples.

```
>>> x = (33)
>>> x
33
>>> y = [33]
>>> y
[33]
```

Reply:
Ah, I see. My reason for replacing some lists with tuples is because of a micro optimization. Since tuples are immutable, the CPython interpreter could optimize the memory usage better. I can change it back to lists if that's preferred?

Reply:
Oh yea, learning to withstand the call for https://wiki.c2.com/?PrematureOptimization is a critical skill. :D

I personally don't like non-functional code changes, because they break blames, and finding stuff like [why the `--exit` option was added`](https://github.com/bleachbit/bleachbit/commit/befe244efee9b2d4859c6b6c31f8bedfd4d85aad#diff-b578cd35e15095f69822ebe497bf8691da1b587d6cc5f5ec252ff4f186dbed56) becomes harder.

Reply:
Boom. 😛

Reply:
I'm quite pleased that I found this bug using pdb, especially since I was able to use the particular cleaner through the CLI. Very recently I was trying to debug a GUI project on github, but the asynchronous event driven nature makes it ever so painful to use pdb or step-by-step debuggers in general. 😢

Reply:
https://github.com/ionelmc/python-hunter is the ultimate debugging tool for me. Just dump all the stuff that is happening and then grep it.

Reply:
I will check it out. Thank you!